wgengine/magicsock/magicsock_linux.go

```diff
@@ -213,6 +213,37 @@ func (c *Conn) receiveDisco(pc net.PacketConn) {
 			// Too small to be a valid UDP datagram, drop.
 			continue
 		}
+
+		dstPort := binary.BigEndian.Uint16(buf[2:4])
+		if dstPort == 0 {
+			c.logf("[unexpected] disco raw: received packet for port 0")
+		}
+
+		var acceptPort uint16
+		if isIPV6 {
+			acceptPort = c.pconn6.Port()
+		} else {
+			acceptPort = c.pconn4.Port()
+		}
+		if acceptPort == 0 {
+			// This should only typically happen if the receiving address family
+			// was recently disabled.
+			c.logf("[v1] disco raw: dropping packet for port %d as acceptPort=0", dstPort)
+			continue
+		}
+
+		// Speculative defense: if dstPort == 0, all disco traffic would be
+		// dropped in conditions such as a host with no ipv6.  This is logged
+		// above, and ignored here so that disco might still flow. If this
+		// condition occurs, it could lead to bugs with behavior such as #5536
+		// with symptoms such as peers only able to communicate disco and no
+		// wireguard data for the duration of a disco session due to a port
+		// mismatch.
```

Comment:
I don't exactly follow this.

Reply:
See the conversation with @danderson above. I tried to reword it a few ways. It's a subtle condition that I believe should never happen, but Dave reports having seen.

I removed the doc string and the dstPort condition as its effectively covered by new conditions above.